x/blockdb/recovery.go

```diff
@@ -0,0 +1,214 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package blockdb
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+
+	"go.uber.org/zap"
+
+	safemath "github.com/ava-labs/avalanchego/utils/math"
+)
+
+// recover detects and recovers unindexed blocks by scanning data files and updating the index.
+// It compares the actual data file sizes on disk with the indexed data size to detect
+// blocks that were written but not properly indexed.
+// For each unindexed block found, it validates the block, then
+// writes the corresponding index entry and updates block height tracking.
+func (s *Database) recover() error {
+	dataFiles, maxIndex, err := s.listDataFiles()
+	if err != nil {
+		return fmt.Errorf("failed to list data files for recovery: %w", err)
+	}
+
+	if len(dataFiles) == 0 {
+		return nil
+	}
+
+	// ensure no data files are missing
+	// If any data files are missing, we would need to recalculate the max height
+	// and max contiguous height. This can be supported in the future but for now
+	// to keep things simple, we will just error if the data files are not as expected.
+	if s.header.MaxDataFileSize > 0 {
+		// Ensure data files are sequential starting from 0
+		for i := 0; i <= maxIndex; i++ {
+			if _, exists := dataFiles[i]; !exists {
+				return fmt.Errorf("%w: data file at index %d is missing", ErrCorrupted, i)
+			}
+		}
+	} else if len(dataFiles) > 1 || maxIndex > 1 {
+		return fmt.Errorf("%w: expect only 1 data file at index 0, got %d files with max index %d", ErrCorrupted, len(dataFiles), maxIndex)
+	}
+
+	// Calculate the expected next write offset based on the data on disk.
+	var calculatedNextDataWriteOffset uint64
+	if s.header.MaxDataFileSize > 0 {
+		// All data files before the last should be full.
```

Comment:
> // All data files before the last should be full.

What if we have written a block that didn't fit into a data file and opened a new one?